extension/observer/ecsobserver/service.go

```diff
@@ -14,12 +14,86 @@
 
 package ecsobserver
 
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"go.uber.org/zap"
+)
+
 type ServiceConfig struct {
 	CommonExporterConfig `mapstructure:",squash" yaml:",inline"`
 
-	// NamePattern is mandetory, empty string means service name based match is skipped.
+	// NamePattern is mandatory.
 	NamePattern string `mapstructure:"name_pattern" yaml:"name_pattern"`
 	// ContainerNamePattern is optional, empty string means all containers in that service would be exported.
 	// Otherwise both service and container name petterns need to metch.
 	ContainerNamePattern string `mapstructure:"container_name_pattern" yaml:"container_name_pattern"`
+
+	// should never be nil because Init must reject it and caller should stop
+	nameRegex *regexp.Regexp
+	// if nil, matches all the container in the task (whose service name is matched by nameRegex)
+	containerNameRegex *regexp.Regexp
+}
+
+func (s *ServiceConfig) Init() error {
```

Comment:
Do these need to be public?

Reply:
Actually this seems a bit weird since we're reusing the config to store non-config fields, I don't think we generally do that. Can you use a different private struct for the initialized config?

Reply:
Renamed it to Validate, and put the initialized value into matcher impl.